### PR TITLE
Make `new AA` allocate the associative array Impl

### PR DESCRIPTION
**Pull request by ntrel**
Fixes Issue 10535.
https://issues.dlang.org/show_bug.cgi?id=10535.

Requires https://github.com/dlang/druntime/pull/3863. See description there for justification.

### CODE REVIEW

**Comment by thewilsonator** [comment]
cc @kinke @ibuclaw this changes glue code

**Comment by thewilsonator** [comment]
Please rebase

**Comment by thewilsonator** [comment]
is this good to go?

**Comment by ntrel** [comment]
@thewilsonator Yes, now marked as ready.

File: compiler/src/dmd/expressionsem.d [hunk 1]
```diff
@@ -3885,6 +3883,15 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
             exp.type = exp.type.pointerTo();
         }
+        else if (tb.ty == Taarray)
+        {
+            // e.g. `new Alias(args)`
+            if (nargs)
+            {
+                exp.error("`new` cannot take arguments for an associative array");
```
Comment (dkorpel) [inline compiler/src/dmd/expressionsem.d:3891]: Please add a test for this

File: test/runnable/newaa.d [hunk 2]
```diff
@@ -0,0 +1,18 @@
+void main()
+{
+	alias AA = int[string];
```
Comment (dkorpel) [inline test/runnable/newaa.d:3]: Use spaces instead of tabs

File: changelog/new-aa.dd [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+This allows two associative array references to point to the same
+associative array instance before any keys have been inserted.
+
+---
+int[string] a = new int[string];
+auto b = a;
+...
+a["seven"] = 7;
+assert(b["seven"] == 7);
```
Comment (dkorpel) [inline changelog/new-aa.dd:11]: This would be a good addition to the runnable test

File: compiler/src/dmd/e2ir.d [hunk 4]
```diff
@@ -1301,6 +1301,16 @@ elem* toElem(Expression e, IRState *irs)
             }
             e = el_combine(ezprefix, e);
         }
+        else if (auto taa = t.isTypeAArray())
```
Comment (dkorpel) [inline compiler/src/dmd/e2ir.d:1304]: I don't understand why Codecov says this isn't covered, it has to be or else the runnable test wouldn't work right?

**Review by dkorpel — APPROVED** [review]

**Pull request merged**